Source/Diagnostics/ReducedDiags/RawEFieldReduction.cpp

```diff
@@ -0,0 +1,122 @@
+/* Copyright 2021 Revathi Jambunathan
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "RawEFieldReduction.H"
+
+#include "Utils/IntervalsParser.H"
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "Utils/WarpXUtil.H"
+
+#include <AMReX_Algorithm.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Vector.H>
+
+#include <algorithm>
+#include <ostream>
+
+#include <regex>
+
+//constructor
+RawEFieldReduction::RawEFieldReduction (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+    using namespace amrex::literals;
+
+    // RZ coordinate is not working
+#if (defined WARPX_DIM_RZ)
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+        "RawEFieldReduction reduced diagnostics does not work for RZ coordinate.");
+#endif
+
+    // read number of levels
+    int nLevel = 0;
+    amrex::ParmParse pp_amr("amr");
+    pp_amr.query("max_level", nLevel);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(nLevel == 0,
+        "RawEFieldReduction reduced diagnostics does not work with mesh refinement.");
+
+    constexpr int noutputs = 3; // Three outputs in the Raw Efield reduction diagnostic (Ex, Ey, Ez)
+    // resize data array
+    m_data.resize(noutputs, 0.0_rt);
+
+    amrex::ParmParse pp_rd_name(rd_name);
+
+    // read reduced function with parser
+    std::string parser_string = "";
+    Store_parserString(pp_rd_name,"reduced_function(x,y,z)", parser_string);
+    m_parser = std::make_unique<amrex::Parser>(makeParser(parser_string,{"x","y","z"}));
+
+    // Replace all newlines and possible following whitespaces with a single whitespace. This
+    // should avoid weird formatting when the string is written in the header of the output file.
+    parser_string = std::regex_replace(parser_string, std::regex("\n\\s*"), " ");
+
+    // read reduction type
+    std::string reduction_type_string;
+    pp_rd_name.get("reduction_type", reduction_type_string);
+    m_reduction_type = GetAlgorithmInteger (pp_rd_name, "reduction_type");
+    if (m_reduction_type == 2) {
+        m_integral_type = GetAlgorithmInteger (pp_rd_name, "integration_type");
+    }
+    amrex::Print() << " reduction type : " << m_reduction_type << " integration type " << m_integral_type << "\n";
+
+    std::string surface_normal_string;
+    pp_rd_name.get("surface_normal", surface_normal_string);
+    if (surface_normal_string == "X") {
+        m_surface_normal[0] = 1;
+    }
+    else if (surface_normal_string == "Y") {
+        m_surface_normal[1] = 1;
+    }
+#if (AMREX_SPACEDIM==3)
+    else if (surface_normal_string == "Z") {
+        m_surface_normal[2] = 1;
+    }
+#endif
```

Comment:
This is a peculiarity of the WarpX code, where in 2D we simulation X-Z .
so I would add the AMREX_SPACEDIM for the y-normal.

Reply:
and not read the y-direction at all (for 2D)

Reply:
Let me check this.